tests/helpers/test_debounce.py

```diff
@@ -95,3 +104,65 @@ async def test_not_immediate_works(hass):
     assert debouncer._timer_task is None
     assert debouncer._execute_at_end_of_timer is False
     debouncer._execute_lock.release()
+    assert debouncer._job.target == debouncer.function
+
+
+async def test_immediate_works_with_function_swapped(hass):
+    """Test immediate works and we can change out the function."""
+    calls = []
+
+    one_function = AsyncMock(side_effect=lambda: calls.append(1))
+    two_function = AsyncMock(side_effect=lambda: calls.append(2))
+
+    debouncer = debounce.Debouncer(
+        hass,
+        None,
+        cooldown=0.01,
+        immediate=True,
+        function=one_function,
+    )
+
+    # Call when nothing happening
+    await debouncer.async_call()
+    assert len(calls) == 1
+    assert debouncer._timer_task is not None
+    assert debouncer._execute_at_end_of_timer is False
+    assert debouncer._job.target == debouncer.function
+
+    # Call when cooldown active setting execute at end to True
+    await debouncer.async_call()
+    assert len(calls) == 1
+    assert debouncer._timer_task is not None
+    assert debouncer._execute_at_end_of_timer is True
+    assert debouncer._job.target == debouncer.function
+
+    # Canceling debounce in cooldown
+    debouncer.async_cancel()
+    assert debouncer._timer_task is None
+    assert debouncer._execute_at_end_of_timer is False
+    assert debouncer._job.target == debouncer.function
+
+    before_job = debouncer._job
+    debouncer.function = two_function
+
+    # Call and let timer run out
+    await debouncer.async_call()
+    assert len(calls) == 2
+    assert calls == [1, 2]
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6309587)